
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Methods &#8212; ChlamDB 2.0 (June 2019) documentation</title>
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script type="text/javascript" src="../_static/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/js/custom.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Website tutorial" href="../tutorial/website.html" />
    <link rel="prev" title="ChlamDB documentation" href="../index.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../tutorial/website.html" title="Website tutorial"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="ChlamDB documentation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ChlamDB 2.0 (June 2019) documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="methods">
<h1>Methods<a class="headerlink" href="#methods" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>A simplified scheme of <code class="docutils literal notranslate"><span class="pre">ChlamDB</span></code> annotation workflow is shown in <strong>Figure 1</strong>. In summary:</p>
<blockquote>
<div><ul class="simple">
<li><p>Annotated genome assemblies integrated into ChlamDB were downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a> (or
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> if the GenBank assembly was not annotated).</p></li>
<li><p>Protein sequences were annotated based on data from multiple source databases and annotation softwares.</p></li>
<li><dl class="simple">
<dt>Protein sequences were clustered into orthologous groups (orthogroups) with <a class="reference external" href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a>.</dt><dd><ul>
<li><p>An alignment and a phylogeny was reconstructed for each orthologous group.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>The closest homologs of each protein were identified in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> and <a class="reference external" href="https://www.uniprot.org/">SwissProt</a></dt><dd><ul>
<li><p>A phylogeny including the closest non-PVC homologs was reconstructed for each orthogroup</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Data were then integrated into a custom SQL database derived from the <a class="reference external" href="https://biosql.org/wiki/Main_Page">bioSQL</a> relational model.</p></li>
</ul>
</div></blockquote>
<div class="align-center figure align-default" id="id2">
<img alt="../_images/workflowV2.png" src="../_images/workflowV2.png" />
<p class="caption"><span class="caption-text"><strong>Figure 1</strong>: Simplified annotation workflow.</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="nextflow-pipeline">
<h2>NextFlow pipeline<a class="headerlink" href="#nextflow-pipeline" title="Permalink to this headline">¶</a></h2>
<p>The annotation pipeline summarized in <strong>Figure 1</strong> was fully automated using <a class="reference external" href="https://www.nextflow.io/">NextFlow</a> .
The code with detailed software versions and command line parameters is available on github <a class="reference external" href="https://github.com/metagenlab/annotation_pipeline_nextflow/blob/master/annotation_pipeline.nf">here</a>.</p>
</div>
<div class="section" id="selection-of-pvc-genomes-integrated-into-the-database">
<h2>Selection of PVC genomes integrated into the database<a class="headerlink" href="#selection-of-pvc-genomes-integrated-into-the-database" title="Permalink to this headline">¶</a></h2>
<p>They are more than 2’000 genome assemblies classified as part of the PVC superphylum (NCBI Taxid 1783257)
on the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&amp;id=1783257&amp;lvl=3&amp;p=gcassembly&amp;lin=f&amp;keep=1&amp;srchmode=1&amp;unlock">NCBI taxonomy website</a>. Those genomes are highly redundant (with a high number of genomes of the
same species, in particular for species of the <em>Chlamydia</em> genus). In order to limit the size of the database,
not all genome assemblies were included into the database.
<code class="docutils literal notranslate"><span class="pre">ChlamDB</span> <span class="pre">version</span> <span class="pre">2.0</span></code> (June 2019) includes:</p>
<blockquote>
<div><ul class="simple">
<li><p>All complete PVC genomes (June 2019).</p></li>
<li><p>All draft <em>Chlamydiae</em> genomes excluding draft genomes from species of
the <em>Chlamydia</em> genus with at least one complete genome (June 2019).</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Draft genomes of the most studied <em>Chlamydia</em> species were excluded because they are highly redundant</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Draft genomes of <em>Planctomycetes</em>, <em>Lentisphaerae</em>, <em>Verrucomicrobia</em> and <em>Kiritimatiellaeota</em> phyla were
excluded to limit the total number of genomes in the database. Additional genomes of clades of interest
might be integrated in future releases of the database.</p>
</div>
<p>The list of genomes is available on <a class="reference external" href="https://chlamdb.ch/#genomes">ChlamDB home page</a>. Complete genomes can be distinguished from draft
genomes based on the number of contigs (“N. contigs” column) since draft assemblies exhibit more than one contig.</p>
</div>
<div class="section" id="protein-annotation">
<h2>Protein annotation<a class="headerlink" href="#protein-annotation" title="Permalink to this headline">¶</a></h2>
<p>Protein sequences were annotated using data from multiple databases and various bioinformatic softwares (<em>Figure 1</em>).
Those annotations are summarized on the “locus” page available for each protein (see for instance the entry of <a class="reference external" href="https://chlamdb.ch/locusx?accession=CT_495">CT_495</a>).</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Transporters</span></code> were annotated using the <a class="reference external" href="http://www.tcdb.org/">Transporter Classification Database</a> (TCDB) and the gBLAST3 software from
the <a class="reference external" href="https://github.com/SaierLaboratory/BioVx">BioV suite</a>. Detailed scores including the <em>E-value</em>, <em>score</em>, <em>percentage of identity</em>, the
<em>number of transmembrane domains</em> identified in the query and the reference protein, the query and hit <em>coverage</em>
are reported on the <a class="reference external" href="https://chlamdb.ch/locusx?accession=CT_495">locus page</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">COG</span></code> annotation was done with <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">rpsblast</a> (with an e-value cutoff of <em>0.001</em>) using <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd_help.shtml#CD_PSSM">position-specific scoring matrix</a>
(PSSM) from the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd.shtml">Conserved Domain Database</a> (CDD)</p></li>
<li><dl class="simple">
<dt>Proteins were assigned to <a class="reference external" href="https://www.genome.jp/kegg/ko.html">Kegg Orthologs</a> (KO) numbers with <a class="reference external" href="ftp://ftp.genome.jp/pub/tools/kofamscan/">KofamScan</a> (default parameters) and <a class="reference external" href="https://www.genome.jp/tools/kofamkoala/">KOfam</a> (a database of Hidden Markov Models [HMM] of <a class="reference external" href="https://www.genome.jp/kegg/ko.html">KEGG Orthologs</a>). KofamScan rely on <a class="reference external" href="http://hmmer.org/">HMMER</a> to search the HMM database.</dt><dd><ul>
<li><p>The <a class="reference external" href="https://www.kegg.jp/kegg/rest/keggapi.html">KEGG API</a> was used to retrieve the annotation of <code class="docutils literal notranslate"><span class="pre">Kegg</span> <span class="pre">Orthologs</span></code> and mapping to <code class="docutils literal notranslate"><span class="pre">KEGG</span> <span class="pre">modules</span></code>, <code class="docutils literal notranslate"><span class="pre">KEGG</span> <span class="pre">pathways</span></code> and <a class="reference external" href="https://www.qmul.ac.uk/sbcs/iubmb/enzyme/">Enzyme Commission numbers</a>.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p><a class="reference external" href="https://www.qmul.ac.uk/sbcs/iubmb/enzyme/">Enzyme Commission numbers</a> were assigned based on <a class="reference external" href="https://www.genome.jp/kegg/ko.html">Kegg Orthologs</a> annotations and with <a class="reference external" href="http://priam.prabi.fr/">PRIAM</a>, a method for automated enzyme detection in fully sequenced genomes. PRIAM uses <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd_help.shtml#CD_PSSM">position-specific scoring matrix</a> (PSSM) of all sequences available in the <a class="reference external" href="https://enzyme.expasy.org/">ENZYME</a> database.</p></li>
<li><dl class="simple">
<dt><a class="reference external" href="https://github.com/ebi-pf-team/interproscan">InterProScan</a> was used to identify protein signatures such as <code class="docutils literal notranslate"><span class="pre">transmembrane</span> <span class="pre">domains</span></code>, <code class="docutils literal notranslate"><span class="pre">Pfam</span> <span class="pre">domains</span></code>, <code class="docutils literal notranslate"><span class="pre">CDD</span> <span class="pre">domains</span></code>,… Detailed InterPro annotations as well as scores associated with each annotation are reported on the “InterproScan” tab of each <a class="reference external" href="https://chlamdb.ch/locusx?accession=CT_495">locus page</a>.</dt><dd><ul>
<li><p><a class="reference external" href="https://github.com/ebi-pf-team/interproscan/wiki/LocalLookupService">The InterProScan match lookup service</a> was used to retrieve pre-calculated InterProScan results for protein integrated into the InterPro database. For sequences not in the lookup service, InterProScan was used to analyse sequences from scratch.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Mapping to <a class="reference external" href="https://www.string-db.org/">STRING</a> (a database of protein-protein interactions) was done by exact protein sequence match.
Protein sequences from SRING were indexed using <a class="reference external" href="https://biopython.org/DIST/docs/api/Bio.SeqUtils.CheckSum-module.html">SEGUID checksums</a>. Associations to the scientific literature (<code class="docutils literal notranslate"><span class="pre">PMID</span></code> tab visible <a class="reference external" href="https://chlamdb.ch/locusx?accession=765098">only if references were found</a>)
were retrived using the <a class="reference external" href="http://version11.string-db.org/help/api/">STRING API</a>.</p></li>
<li><p>Protein subcellular localization predictions were done using <a class="reference external" href="https://www.psort.org/psortb/index.html">PSORTb</a> (withe the <code class="docutils literal notranslate"><span class="pre">--negative</span></code> argument).</p></li>
<li><dl class="simple">
<dt>Effectors of the type III secretion system were predicted using four different softwares:</dt><dd><ul>
<li><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/27/6/777/235789">BPBAac</a> (default parameters)</p></li>
<li><p><a class="reference external" href="https://academic.oup.com/nar/article/39/suppl_1/D591/2507337">effectiveT3</a> (model <code class="docutils literal notranslate"><span class="pre">TTSS_STD-2.0.2</span></code>, cutoff of <code class="docutils literal notranslate"><span class="pre">0.9999</span></code>)</p></li>
<li><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/35/12/2051/5165378">DeepT3</a>  (default parameters)</p></li>
<li><p><a class="reference external" href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0058173">T3_MM</a> (default parameters)</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>TODO: UNIPARC - SwissProt</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="homology-search-refseq-and-swissprot">
<h2><span class="target" id="homology-search">Homology search</span> (RefSeq and SwissProt)<a class="headerlink" href="#homology-search-refseq-and-swissprot" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>The closest identifiable <a class="reference external" href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> homologs were searched with <a class="reference external" href="https://www.nature.com/articles/nmeth.3176">Diamond</a> (parameters: <code class="docutils literal notranslate"><span class="pre">--max-target-seqs</span></code> 200 <code class="docutils literal notranslate"><span class="pre">-e</span></code> 0.01 <code class="docutils literal notranslate"><span class="pre">--max-hsps</span></code> 1). The 200 first hits were retained for each protein.</p></li>
<li><p>The closest identifiable <a class="reference external" href="https://www.uniprot.org/">SwissProt</a> homologs were searched with BLASTp (parameters: <code class="docutils literal notranslate"><span class="pre">-evalue</span></code> 0.001). The 100 first hits were retained for each protein.</p></li>
</ul>
</div>
<div class="section" id="id1">
<h2><span class="target" id="identification-of-orthogroups">Identification of Orthogroups</span><a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>Orthogroups (or orthologous groups) were identified with <a class="reference external" href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a>. This tools identify orthogroups based on BLASTp (parameters: <code class="docutils literal notranslate"><span class="pre">-evalue</span></code> 0.001)
results using the <a class="reference external" href="https://micans.org/mcl/">MCL</a> clustering software.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code class="docutils literal notranslate"><span class="pre">Orthologs</span></code> are pairs of genes that descended from a single gene in the last common ancestor (LCA) of two species.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An <code class="docutils literal notranslate"><span class="pre">orthogroup</span></code> is the group of genes descended from a single gene in the last common ancestor (LCA) of a group of species.
As gene duplication and loss occur frequently in bacteria, we rarely have exactly one ortholog in each considered genome.</p>
</div>
</div>
<div class="section" id="orthogroup-multiple-sequence-alignments-phylogenies">
<h2>Orthogroup multiple sequence alignments &amp; phylogenies<a class="headerlink" href="#orthogroup-multiple-sequence-alignments-phylogenies" title="Permalink to this headline">¶</a></h2>
<p>Protein sequences of each orthogroup were aligned with <a class="reference external" href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> (default parameters). A phylogeny was then reconstructed for each orthogroup of three or more sequences. The phylogeny was reconstructed with <a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a> with default parameters. The node support values are not traditionnal <a class="reference external" href="https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1558-5646.1985.tb00420.x">boostrap support values</a>. <a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a> uses the Shimodaira-Hasegawa test with 1,000 bootstrap replicates to quickly estimate the reliability of each split in the tree. Values higher than <strong>0.95</strong> can be considered as “strongly supported”.</p>
</div>
<div class="section" id="phlogeny-including-top-refseq-hits">
<h2>Phlogeny including top RefSeq hits<a class="headerlink" href="#phlogeny-including-top-refseq-hits" title="Permalink to this headline">¶</a></h2>
<p>A second phylogeny was reconstructed for each orthogroup. This phylogeny includes the 4 best  non-PVC RefSeq hits of each protein (see the <a class="reference internal" href="#homology-search">Homology search</a> paragraph).
First, the NCBI taxon ID of each RefSeq hit was retrieved using the <code class="docutils literal notranslate"><span class="pre">prot.accession2taxid</span></code> mapping file available from the <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/">NCBI taxonomy ftp website</a>. PVC hits were removed and the amino acid sequence of the 4 best non-PVC hits was retrieved from the NCBI using the <a class="reference external" href="https://biopython.org/DIST/docs/api/Bio.Entrez-module.html">biopython interface to Entrez</a>. Protein sequences of each orthogroup + RefSeq homologs were aligned with <a class="reference external" href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> (default parameters) and the phylogeny was reconstructed with <a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a> (default parameters).</p>
<p></p>
<p>This phylogeny allows users to check whether PVC proteins form a monophyletic group or if they are for instance hints of horizontal gene transfer(s) with bacteria from other phyla.</p>
</div>
<div class="section" id="calculation-of-pairwise-protein-identity">
<h2>Calculation of pairwise protein identity<a class="headerlink" href="#calculation-of-pairwise-protein-identity" title="Permalink to this headline">¶</a></h2>
<p>Pairwise protein sequence identities reported on ChlamDB were calculated based on the multiple sequence alignments of orthologous groups (two see previous paragraph). The identity between two sequences can be calculated in different ways (see <a class="reference external" href="http://lh3.github.io/2018/11/25/on-the-definition-of-sequence-identity">this blog post discussing</a> that topic). We calculated the identity by excluding gaps and calculating the identity based on aligned positions only:</p>
<ul class="simple">
<li><p>number of matches / ( number of matches +  number of mismatches)</p></li>
</ul>
<p>Nevertheless, if the alignment covered less than 30% of one of the two compared sequences, the identity was set to 0.</p>
</div>
<div class="section" id="circular-genome-plots-circos">
<h2>Circular genome plots (Circos)<a class="headerlink" href="#circular-genome-plots-circos" title="Permalink to this headline">¶</a></h2>
<p>Circular genome plots are generated dynamically with <a class="reference external" href="http://circos.ca/">Circos</a>. These plots are not generated based on the alignment of DNA sequences but based on orthology data (see previous paragraphs). The two outer gray circles show the location of <a class="reference external" href="https://en.wikipedia.org/wiki/Open_reading_frame">open reading frames</a> (ORFs) encoded on the leading and lagging strand of the reference genome (<strong>Figure 2 A and B</strong>). The red/blue inner circles show the conservation of each protein encoding ORF in one or multiple other genomes. Identity values were pre-calculated from protein alignments (see previous paragraphs). If the compared genomes encode more than one ortholog, the highest identity is used to color the region.</p>
<div class="align-center figure align-default" id="id3">
<a class="reference internal image-reference" href="../_images/circos_method.png"><img alt="../_images/circos_method.png" src="../_images/circos_method.png" style="width: 100%;" /></a>
<p class="caption"><span class="caption-text"><strong>Figure 2</strong>: Example of circular genome plot. <strong>A)</strong> Compared genomes are ordered based on the median protein identity with the reference genome. <strong>B)</strong> Zoom showing the detail of a genomic region. All circular plots are interactive and users can click on any ORF to access the detailed annotation page of the corresponding protein.</span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="species-phylogeny">
<h2>Species phylogeny<a class="headerlink" href="#species-phylogeny" title="Permalink to this headline">¶</a></h2>
<p>The reference phylogeny was reconstructed with <a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a> (default parameters, JTT+CAT model) based on the concatenated alignment of 32 single copy orthogroups conserved in at least 266 out of the 277 genomes part of <code class="docutils literal notranslate"><span class="pre">ChlamDB</span> <span class="pre">2.0</span></code>.</p>
</div>
<div class="section" id="comparative-analyses-of-orthogroups-and-cog-kegg-interpro-annotations">
<h2>Comparative analyses of orthogroups and COG/KEGG/Interpro annotations<a class="headerlink" href="#comparative-analyses-of-orthogroups-and-cog-kegg-interpro-annotations" title="Permalink to this headline">¶</a></h2>
<p>Orthogroups KEGG, COG, Interpro entries</p>
</div>
<div class="section" id="prediction-of-protein-protein-interactions">
<h2>Prediction of protein-protein interactions<a class="headerlink" href="#prediction-of-protein-protein-interactions" title="Permalink to this headline">¶</a></h2>
<p>The interactions reported in the “interactions” tab were predicted based on two different approaches: <code class="docutils literal notranslate"><span class="pre">phylogenetic</span> <span class="pre">profiling</span></code> and <code class="docutils literal notranslate"><span class="pre">conservation</span> <span class="pre">of</span> <span class="pre">gene</span> <span class="pre">neighborhood</span></code>. Indeed, proteins that physically interact tend to be encoded in a close neighborhood and tend to co-occur in the same genomes (see <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/9787636">Dandekar  et al</a> and <a class="reference external" href="https://royalsocietypublishing.org/doi/10.1098/rsif.2007.1047">Kensche et al.</a> for more details).</p>
<div class="section" id="phylogenetic-profiling">
<h3>1 Phylogenetic profiling<a class="headerlink" href="#phylogenetic-profiling" title="Permalink to this headline">¶</a></h3>
<p>Orthogroup exhibiting similar patterns of presence/absence were identified by calculating the euclidian and jaccard distances of all pairs of orthologroups phylogenetic profiles (see <strong>Figure 3</strong>). See <a class="reference external" href="https://royalsocietypublishing.org/doi/10.1098/rsif.2007.1047">Kensche et al.</a> for a review on phylogenetic profiling methods.</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Phylogenetic profiles were collapsed at the species level. If an orthogroup was present in only a subset of the strains of the considered species, it was still considered as present in that species.</p></li>
<li><p>Pairwise <a class="reference external" href="https://en.wikipedia.org/wiki/Euclidean_distance">euclidean</a> and <a class="reference external" href="https://en.wikipedia.org/wiki/Jaccard_index">jaccard</a> distances were calculated between pairs of profiles</p></li>
<li><p>The default euclidian distance cutoff to report interactions is <code class="docutils literal notranslate"><span class="pre">2.2</span></code>. If more than 30 profiles had an euclidian distance smaller or equal to <code class="docutils literal notranslate"><span class="pre">2.2</span></code>, the stringency was adjusted incrementally with cutoffs of <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code> and finally <code class="docutils literal notranslate"><span class="pre">0</span></code>. If more than 30 profiles exhibit the exact same profile, nothing is reported. Indeed, it would by typically poorly informative profiles (e.g. proteins conserved in all species).</p></li>
</ol>
</div></blockquote>
<div class="align-center figure align-default" id="id4">
<a class="reference internal image-reference" href="../_images/profile.svg"><img alt="../_images/profile.svg" src="../_images/profile.svg" width="500px" /></a>
<p class="caption"><span class="caption-text"><strong>Figure 3</strong>: Schematic view of the phylogenetic profiling method used to predict protein-protein interactions. Distances between profiles of presence/absence of orthologs were calculated using two different metrics (<a class="reference external" href="https://en.wikipedia.org/wiki/Euclidean_distance">euclidean</a> and <a class="reference external" href="https://en.wikipedia.org/wiki/Jaccard_index">jaccard</a> distances).</span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Have a look at the predicted interactors of the <a class="reference external" href="https://chlamdb.ch/profile_interactions/group_414">cell shape-determining protein MreB</a>. The detailed profile of each orthogroup is available in the “profile” tab.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The profiles 5 and 6 in <strong>Figure 3</strong> are strictly identitcal but poorly informative since orthologs were identified in all considered species. Spurious associations like this one cannot always be automatically discarded. Thus, those associations should always be interpreted with caution. The detailed profile of associated orthogroups is always provided (“profile” tab).</p>
</div>
</div>
<div class="section" id="identification-of-conserved-gene-neighborhood-in-different-species">
<h3>2. Identification of conserved gene neighborhood in different species<a class="headerlink" href="#identification-of-conserved-gene-neighborhood-in-different-species" title="Permalink to this headline">¶</a></h3>
<p>Genes encoded in a close neighborhood in distantly related species were identified based on orthology data (see <a class="reference internal" href="#identification-of-orthogroups">Identification of Orthogroups</a>). See for example the case of the subunits of the <a class="reference external" href="https://www.chlamdb.ch/neig_interactions/wcw_1123">F-type ATPase</a> that are systematically encoded in a close neighborhood. The method was the following:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>protein encoding genes were iterated for each genome</p></li>
<li><p>if the considered protein encoding gene had one (or multiple) homolog(s) in distantly related genomes (a cutoff of 60% median protein identity was used to consider two genomes sufficently distant)</p></li>
<li><p>the 10 kilobases upstream and downstream of each ortholog were compared to the reference genome to identify conserved  neighbors (see <strong>Figure 4.A</strong>)</p></li>
<li><p>the ratio of co-occurence of pairs of orthogroups was used to score the association between pairs of locus (<strong>Figure 4.B</strong>)</p></li>
<li><p>only proteins exhibiting a <code class="docutils literal notranslate"><span class="pre">conservation</span> <span class="pre">score</span> <span class="pre">of</span> <span class="pre">0.8</span></code> are reported in the “interactions” tab of locus pages</p></li>
</ol>
</div></blockquote>
<div class="align-center figure align-default" id="id5">
<a class="reference internal image-reference" href="../_images/conserved_neig.svg"><img alt="../_images/conserved_neig.svg" src="../_images/conserved_neig.svg" width="100%" /></a>
<p class="caption"><span class="caption-text"><strong>Figure 4</strong>: Identification of conserved gene neighborhood. <strong>A)</strong> Example with the red locus, part of the orthologous group <code class="docutils literal notranslate"><span class="pre">group_222</span></code>. This locus has orthologs in 3 other genomes.
Proteins encoded 10 kilobases upstream and 10 kilobases downstream of the reference locus are extracted and compared between the reference
genome and all other genomes in which an ortholog is present. <strong>B)</strong> The <code class="docutils literal notranslate"><span class="pre">conservation</span> <span class="pre">score</span></code> of a pair of locus is the ratio of cases for which orthologs were encoded in the green windows out of the total number of comparisons. For instance, the green locus is encoded less than 10kb upstream of the red locus in two genomes, but more than 10kb appart in the third genome. Is score is thus of 2/3 (2 out of 2 comparsions). The pink locus is always encoded in the vicinity of the red locus in all compared genomes. Its score is thus of 3/3. <strong>C)</strong> If multiple orthologs were identified if the compared genome, only the most similar one is considered for the comparison (based on amino acid identity calculated based on the orthogroup multiple sequence alignment).</span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Since orthologous groups identified by <a class="reference external" href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a> include both orthologs and paralogs, the relationship between pair of genomes is not necessarily 1 to 1 (have a look at the <a class="reference external" href="https://omabrowser.org/oma/type/">OMA website</a> to read detailed explanations of the different types of orthologs). If multiple orthologs were encoded in the compared genome, only the most similar one was used for the comparison (<strong>Figure 4.C</strong>). The choice of the closest locus was based on protein identity and not based on phylogenetic trees, which means that we might not always consider the most closely related protein encoded in the compared genome for the comparison.</p>
</div>
</div>
</div>
<div class="section" id="prediction-of-candidate-type-iii-secretion-system-effectors">
<h2>Prediction of candidate type III secretion system effectors<a class="headerlink" href="#prediction-of-candidate-type-iii-secretion-system-effectors" title="Permalink to this headline">¶</a></h2>
<p>Effectors of the type III secretion system are poorly conserved. In addition, the signal allowing to specifically secrete effector proteins is not clearly identified. It is thus difficult to identify effectors in newly sequenced genomes. Predition methods generally rely on the use classifiers that are trained on a set of known effector proteins. Since all <em>Chlamydiae</em> genomes encode a type III secretion system, candidate effectors were identified using four different classifiers:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/27/6/777/235789">BPBAac</a> (default parameters)</p></li>
<li><p><a class="reference external" href="https://academic.oup.com/nar/article/39/suppl_1/D591/2507337">effectiveT3</a> (model <code class="docutils literal notranslate"><span class="pre">TTSS_STD-2.0.2</span></code>, cutoff of <code class="docutils literal notranslate"><span class="pre">0.9999</span></code>)</p></li>
<li><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/35/12/2051/5165378">DeepT3</a>  (default parameters)</p></li>
<li><p><a class="reference external" href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0058173">T3_MM</a> (default parameters)</p></li>
</ul>
</div></blockquote>
<p>Results of all four tools are reported on the <a class="reference external" href="https://chlamdb.ch/locusx?accession=CT_495">locus page</a>  of each protein (in the <code class="docutils literal notranslate"><span class="pre">T3SS</span> <span class="pre">effectors</span> <span class="pre">prediction</span></code> section).</p>
</div>
<div class="section" id="taxonomic-profile-of-pfam-domains-and-cogs">
<h2>Taxonomic profile of Pfam domains and COGs<a class="headerlink" href="#taxonomic-profile-of-pfam-domains-and-cogs" title="Permalink to this headline">¶</a></h2>
<p>Coding sequences of the 6661 reference and representative genomes available from RefSeq (September 2017) were downloaded from RefSeq ftp (<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/refseq/">ftp://ftp.ncbi.nlm.nih.gov/refseq/</a>). Hidden Markov models of the PFAM database (6) were used to search for sporulation-related domains using hmmsearch (HMMER version 3.1b2, (7)). Hits were filtered based on PFAM trusted cutoffs. Genomes were then grouped by order based on the NCBI Taxonomy database (8) with the python library ete2. Only orders exhibiting a minimum of 5 genomes are reported.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://chlamdb.ch/pfam_profile/PF01823/phylum">https://chlamdb.ch/pfam_profile/PF01823/phylum</a></p></li>
</ul>
<div class="align-center figure align-default">
<a class="reference internal image-reference" href="methods/../img/PF08486_phylum_profile.svg"><img alt="methods/../img/PF08486_phylum_profile.svg" src="methods/../img/PF08486_phylum_profile.svg" width="600px" /></a>
</div>
</div>
<div class="section" id="evaluation-of-the-quality-and-completeness-of-draft-genomes">
<h2>Evaluation of the quality and completeness of draft genomes<a class="headerlink" href="#evaluation-of-the-quality-and-completeness-of-draft-genomes" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="protein-accessions-mapping">
<h2>Protein accessions mapping<a class="headerlink" href="#protein-accessions-mapping" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>RefSeq, GenBank, UniProt</p></li>
</ul>
</div>
<div class="section" id="sql-database-and-web-interface">
<h2>SQL database and Web Interface<a class="headerlink" href="#sql-database-and-web-interface" title="Permalink to this headline">¶</a></h2>
<p>TODO
* database size 80Gb
* Figure generated from stores data</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Thee SQL database is about 80GB in size. Protein alignments are not stored into the database.</p>
</div>
</div>
<div class="section" id="source-databases">
<h2>Source databases<a class="headerlink" href="#source-databases" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-center" id="id6">
<caption><span class="caption-text">Version of source databases used for the annotation</span><a class="headerlink" href="#id6" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 45%" />
<col style="width: 55%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Database name</p></th>
<th class="head"><p>Version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>UniprotKB-UNIPARC</p></td>
<td><p>2019.03</p></td>
</tr>
<tr class="row-odd"><td><p>TCDB</p></td>
<td><p>2019.06</p></td>
</tr>
<tr class="row-even"><td><p>RefSeq</p></td>
<td><p>90</p></td>
</tr>
<tr class="row-odd"><td><p>CDD (COG)</p></td>
<td><p>v3.17</p></td>
</tr>
<tr class="row-even"><td><p>InterProScan</p></td>
<td><p>5.35-74.0</p></td>
</tr>
<tr class="row-odd"><td><p>STRING</p></td>
<td><p>v11</p></td>
</tr>
<tr class="row-even"><td><p>PDB</p></td>
<td><p>2019.06</p></td>
</tr>
<tr class="row-odd"><td><p>COG/CDD</p></td>
<td><p>v3.17</p></td>
</tr>
<tr class="row-even"><td><p>KoFam</p></td>
<td><p>2019.04.09</p></td>
</tr>
<tr class="row-odd"><td><p>PRIAM</p></td>
<td><p>2018.06</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="software-versions">
<h2>Software versions<a class="headerlink" href="#software-versions" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-center" id="id7">
<caption><span class="caption-text">Version of the main softwares used for protein annotation</span><a class="headerlink" href="#id7" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 52%" />
<col style="width: 48%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Software name</p></th>
<th class="head"><p>Version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>FastTree</p></td>
<td><p>2.1.10</p></td>
</tr>
<tr class="row-odd"><td><p>Diamond</p></td>
<td><p>0.9.24</p></td>
</tr>
<tr class="row-even"><td><p>OrthoFinder</p></td>
<td><p>2.2.7</p></td>
</tr>
<tr class="row-odd"><td><p>BLAST</p></td>
<td><p>2.7.1</p></td>
</tr>
<tr class="row-even"><td><p>CheckM</p></td>
<td><p>1.0.12</p></td>
</tr>
<tr class="row-odd"><td><p>KoFamScan</p></td>
<td><p>2019/4/9</p></td>
</tr>
<tr class="row-even"><td><p>Mafft</p></td>
<td><p>7.407</p></td>
</tr>
<tr class="row-odd"><td><p>PSORTb</p></td>
<td><p>3.0.6</p></td>
</tr>
<tr class="row-even"><td><p>HMMER</p></td>
<td><p>3.1b2</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="code-availability">
<h2>Code availability<a class="headerlink" href="#code-availability" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 39%" />
<col style="width: 61%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Website interface</p></td>
<td><p><a class="reference external" href="https://github.com/metagenlab/chlamdb">https://github.com/metagenlab/chlamdb</a></p></td>
</tr>
<tr class="row-even"><td><p>Annotation pipeline</p></td>
<td><p><a class="reference external" href="https://github.com/metagenlab/annotation_pipeline_nextflow">https://github.com/metagenlab/annotation_pipeline_nextflow</a></p></td>
</tr>
<tr class="row-odd"><td><p>Public database download and indexing</p></td>
<td><p><a class="reference external" href="https://github.com/metagenlab/annotation_pipeline_snakemake">https://github.com/metagenlab/annotation_pipeline_snakemake</a></p></td>
</tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Methods</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#nextflow-pipeline">NextFlow pipeline</a></li>
<li><a class="reference internal" href="#selection-of-pvc-genomes-integrated-into-the-database">Selection of PVC genomes integrated into the database</a></li>
<li><a class="reference internal" href="#protein-annotation">Protein annotation</a></li>
<li><a class="reference internal" href="#homology-search-refseq-and-swissprot">Homology search (RefSeq and SwissProt)</a></li>
<li><a class="reference internal" href="#id1">Identification of Orthogroups</a></li>
<li><a class="reference internal" href="#orthogroup-multiple-sequence-alignments-phylogenies">Orthogroup multiple sequence alignments &amp; phylogenies</a></li>
<li><a class="reference internal" href="#phlogeny-including-top-refseq-hits">Phlogeny including top RefSeq hits</a></li>
<li><a class="reference internal" href="#calculation-of-pairwise-protein-identity">Calculation of pairwise protein identity</a></li>
<li><a class="reference internal" href="#circular-genome-plots-circos">Circular genome plots (Circos)</a></li>
<li><a class="reference internal" href="#species-phylogeny">Species phylogeny</a></li>
<li><a class="reference internal" href="#comparative-analyses-of-orthogroups-and-cog-kegg-interpro-annotations">Comparative analyses of orthogroups and COG/KEGG/Interpro annotations</a></li>
<li><a class="reference internal" href="#prediction-of-protein-protein-interactions">Prediction of protein-protein interactions</a><ul>
<li><a class="reference internal" href="#phylogenetic-profiling">1 Phylogenetic profiling</a></li>
<li><a class="reference internal" href="#identification-of-conserved-gene-neighborhood-in-different-species">2. Identification of conserved gene neighborhood in different species</a></li>
</ul>
</li>
<li><a class="reference internal" href="#prediction-of-candidate-type-iii-secretion-system-effectors">Prediction of candidate type III secretion system effectors</a></li>
<li><a class="reference internal" href="#taxonomic-profile-of-pfam-domains-and-cogs">Taxonomic profile of Pfam domains and COGs</a></li>
<li><a class="reference internal" href="#evaluation-of-the-quality-and-completeness-of-draft-genomes">Evaluation of the quality and completeness of draft genomes</a></li>
<li><a class="reference internal" href="#protein-accessions-mapping">Protein accessions mapping</a></li>
<li><a class="reference internal" href="#sql-database-and-web-interface">SQL database and Web Interface</a></li>
<li><a class="reference internal" href="#source-databases">Source databases</a></li>
<li><a class="reference internal" href="#software-versions">Software versions</a></li>
<li><a class="reference internal" href="#code-availability">Code availability</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../index.html"
                        title="previous chapter">ChlamDB documentation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../tutorial/website.html"
                        title="next chapter">Website tutorial</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/methods/annotation.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../tutorial/website.html" title="Website tutorial"
             >next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="ChlamDB documentation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ChlamDB 2.0 (June 2019) documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Trestan Pillonel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.
    </div>
  </body>
</html>