


<!DOCTYPE html> <html>  <head> {% load custom_tags %}
      {% load static %}
      {% load static %}
      {% load crispy_forms_tags %}
      {% include "chlamdb/header.html" %}
  </head>
 
  <div class="container-fluid" id="main_container">
    <div class="row">
      <div id="wrapper">
        <div id="page-content-wrapper">
          <div class="row">
            <div class="col-lg-12">
             {% include "chlamdb/menu.html" %}
             <div class="home-title"><b> Orthogroups </b></div>
             <hr class="lines-home">

              <nav>
                  <ul id="tabs_main" class="nav nav-tabs">
                      <li class="active"><a href="{% url 'extract_orthogroup' %}">Detailed comparison</a></li>
                      <li><a href="{% url 'venn_orthogroup' %}">Venn diagram</a></li>
                      <li><a href="{% url 'orthogroup_comparison' %}">Compare Orthogroups size</a></li>
                      <li><a href="{% url 'plot_heatmap' 'orthology' %}">Whole proteomes heatmaps</a></li>
                      <li><a href="{% url 'pan_genome' 'orthology' %}">Pan/Core genome plots</a></li>
                  </ul>
              </nav>

             <div id="form" class="col-lg-10 col-md-12 col-sm-12">


                <div class="panel panel-success" style="width:100%; top: 200px; margin: 10px 10px 10px 10px">
                    <div class="panel-heading" style="width:100%">
                        <h3 class="panel-title">Help</h3>
                    </div>
                    <p style="margin: 10px 10px 10px 10px; line-height: 180%">Get list of proteins shared between genomes selected in panel A and absent in genomes of panel B. Clustering based on BLAST usting orthoFinder.
                    Reference: one of the genome in panel A can be chosen as reference for display the result orthogroup list on a circular map of the genome. Click on "show on circular map" to generate the circos map. Corresponding locus are highlighted in pink. <br>
                    Missing data: if more than one panel A genome (include) is selected, possibility to tolerate missing data in 1 up to 10 genomes (in order to get orthogroup present in at least x out of the y genomes selected in panel A).</p>
                </div>
                <!--
                <div style="position: absolute;margin-top:10%; margin-left: 33%;">
                    <div style="height:40px;margin-left: 180px;width:280px;border: 2px solid #8AC007; border-radius: 15px;">
                        <div style="width: 100%; float: left; margin-left: 10px;margin-top:3px">
                            <p>Consider single copy groups only</p>
                        </div >

                        <div style="height: 40px; margin-left: 180px;margin-top:3px;">
                            <label class="switch-light well" onclick="" style="margin: 5px 0px 0px 0px; padding: 0 0 0 0; position: absolute">
                                <input type="checkbox" style="padding: 0 0px 0 0" id="button_single_copy" name="button_single_copy">
                                <span>
                                <span>Off</span>
                                <span>On</span>
                                </span>
                                <a class="btn btn-success btn-xs"></a>
                            </label>
                        </div>
                    </div>
                </div>
                -->
                {% block content %}
                <form action='{% url "extract_orthogroup" %}' method="post" id="extract-form">
                    {% csrf_token %}
                    
                    {% crispy form %}
                </form>
                {% endblock %}
            </div>

            {% if wrong_n_missing%}

            <div class="panel panel-warning" style="width:500px ; top: 200px; margin: 10px 10px 10px 10px">
                <div class="panel-heading" style="width:100%">
                    <h3 class="panel-title">Help</h3>
                </div>
                <p style="margin: 10px 10px 10px 10px">You cannot set a number of missing data bigger than the number of included genomes</p>
            </div>

            {% elif no_match%}

            <div class="panel panel-warning" style="width:500px ; top: 200px; margin: 10px 10px 10px 10px">
                <div class="panel-heading" style="width:100%">
                    <h3 class="panel-title"></h3>
                </div>
                <p style="margin: 10px 10px 10px 10px">No match</p>
            </div>

            {% endif %}

			{% if envoi_extract %}

			<div class="row" style="padding-top:20px">
			<div class="col-lg-12">
			  <ul id="tabs" class="nav nav-tabs" data-tabs="tabs">
				  <li class="active"><a href="#red" data-toggle="tab">Orthogroups</a></li>
				  <li><a href="#orange" data-toggle="tab">Table detail</a></li>
			  </ul>

			  <div id="my-tab-content" class="tab-content">
				  <div class="tab-pane active" id="red">
					<div style="padding-top:10px;">

						<h3>Number of orthogroups: {{ sum_group }} </h3>


						  <div class="panel panel-success" style="width:100%; top: 200px; margin: 10px 10px 10px 10px">
							  <div class="panel-heading" style="width:100%">
								  <h3 class="panel-title">Help</h3>
							  </div>
							  <p style="margin: 10px 10px 10px 10px; line-height: 180%">The annotation(s) of orthologous groups is a consensus of the annotation of all members of the group. For each annotation (gene name, product, COG,...), 
							  only the two most frequent annotations are reported. Number in parentheses indicates the number of occurences of the annotation (how many proteins of the orthologous group have this exact annotation). </br>
							  The second tab reports the complete list of protein of reference genome(s).                               
							  </p>
						  </div>


						<table class="hover" id="table_groups">

							<thead>
								<tr>
								{% for value in table_headers %}
									<th> {{value|safe}} </th>
								{% endfor %}
								</tr>
							</thead>

							<tbody>
							{% for value in match_groups_data %}
								<tr>
									{% for entry in value %}
									{% if forloop.first %}
										<td><a href="{% url 'orthogroup' entry %}">{{entry}}</a></td>
									{% else %}
										<td>{{entry|safe}}</td>
									{% endif %}
									{% endfor %}
								</tr>
							{% endfor %}
							</tbody>
						</table>
					</div>
				  </div>
				  <div class="tab-pane" id="orange">
						  <div id="export_bouttons_groups">
							  {% if show_reference_annot %}
							  <button type="button" class="btn btn-primary btn-xs" onclick="location.href='/chlamdb/fasta/{{fasta_ref_url|safe}}'">fasta aa</button>
							  {% endif %}
								<br/>
						  </div>


						  <h3>Number of orthogroups: {{ sum_group }} </h3>

						  <table class="hover" id="orthogroups_table">
							  <thead>
								  <tr>
								  {% for value in details_header %}
									<th style="text-align:center; padding: 5px 30px"> {{value|safe}} </th>
								  {% endfor %}
								  </tr>
							  </thead>

							  <tbody>

								  {% for entry in details_data %}
									  <tr>
										  {% for values in entry %}
										  <td style="text-align: center;"> {{ values|safe }} </td>
										  {% endfor %}
									  </tr>
								  {% endfor %}
							  </tbody>
						  </table>
				  </div>
			  </div>
        </div>
		{% endif %}
        
        </div>

        </div> <!-- row -->
      </div> <!-- page-content-wrapper -->
    </div> <!-- wrapper -->
  </div>  <!-- row -->
  
</div>  <!-- container-fluid -->

<script>

  $(document).ready(function() {
	  $('#orthogroups_table').DataTable();
	  $("#table_groups").DataTable();
  } );

</script>

{% include "chlamdb/style_menu.html" %}
{% include "chlamdb/show_hide_plasmid_accessions.html" %}
</body>
</html>
