<!DOCTYPE html>


<html>

<head>

<meta name="description" content="A comparative genomics database of the phylum Chlamydiae and other PVC (Planctomycetes, Verrucomicrobia, Lentisphaerae).">

    {% load static %}
    {% load static %}

{% include "chlamdb/header.html" %}

</head>

<body>
    <div class="container-fluid" id="main_container">
      <div class="row">
        <div id="wrapper">
            <div id="page-content-wrapper">
              <div class="row">
                <div class="col-lg-12">
                   {% include "chlamdb/menu.html" %}
                    <ul id="tabs" class="nav nav-tabs" data-tabs="tabs">
                        <li class="active"><a href="#home" data-toggle="tab">Home</a></li>
                        <li><a href="#genomes" data-toggle="tab">Genomes</a></li>
                        <li><a href="#phylogeny" data-toggle="tab">Phylogeny</a></li>
                    </ul>


                    <div id="my-tab-content" class="tab-content">
                        <div class="tab-pane active" id="home">


                            <div style="padding-top:20px;">
                                <div class="jumbotron">
                                <h1>Chlamydiae Database </h1>
                                <center>
                                  <p>
                                  A comparative genomics database of the  <i>Planctomycetes</i>-<i>Verrucomicrobia</i>-<i>Chlamydiae</i> superphylum
                                  </p>
                                </center>


                                <!--<a id="learn_more" class="btn btn-success btn-lg" href="{% url 'help' %}" role="button">Learn more</a>-->

                                </div>
                            </div>

                            {% if missing_mandatory|length != 0%}
                                <div class="alert alert-danger" role="alert">
                                <h4 class="alert-heading">Warning: incomplete setup</h4>
                                <p>The following data is missing from the database: 
                                
                                      <ol>
                                      {% for item in missing_mandatory %}
                                          <li>{{item}}</li>
                                      {% endfor %}
                                      
                                      
                                      </ol>
                                  This database misght not be fully functionnal

                                </p>
                              </div>
                            {% endif %}


                            <div class="row">
                              <div class="col-lg-6">


                                <div class="panel panel-default">
                                  <div class="panel-heading">
                                    <h3 class="panel-title">Content of the database & main features</h3>
                                  </div>
                                  <div class="panel-body" style="margin: 10px 10px 10px 10px">

                                   <font size="2">
                                    <ul style="list-style-type:square;">

                                      {% if biodb == "chlamydia_04_16"%}

                                       <li><a href="/home/#genomes">{{n_genomes}} <i>Chlamydiae</i> genomes from 26 different species</a> (see the <a href="/home#phylogeny">phylogeny</a>)</li>

                                      {% elif biodb == "2019_06_PVC"%}
                                         <li><a href="/home/#genomes">{{n_genomes}} <i>PVC</i> genomes</a> (see the <a href="/home/#phylogeny">phylogeny</a>)</li>
                                      {% else %}
                                          <li><a href="/home/#genomes">{{n_genomes}} <i>PVC</i> genomes</a> (see the <a href="/home/#phylogeny">phylogeny</a>)</li>
                                      {% endif %}

                                       <li>Orthologous proteins identified with <a href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a></li>
                                       <li>Detailed annotation of proteins based on</li>
                                           <ul style="list-style-type:circle; padding-left:20px;">
                                             <li><a href="https://www.genome.jp/kegg/">KEGG</a> (Kegg orthologs)</li>
                                             <li><a href="https://www.ncbi.nlm.nih.gov/COG/">COG</a> (clusters of orthologous groups)</li>
                                             <li><a href="https://string-db.org/">STRING</a> (protein-protein interactions)</li>
                                             <li><a href="http://priam.prabi.fr/">PRIAM</a> (enzymes)</li>
                                             <li><a href="http://www.tcdb.org/">TCDB</a> (curated transporters annotation)</li>
                                             <li><a href="https://www.uniprot.org/">SwissProt</a> (manually annotated proteins)</li>
                                             <li><a href="https://www.ebi.ac.uk/interpro/">InterproScan</a> (domains and signatures)</li>
                                             <li><a href="http://csbl.bmb.uga.edu/DOOR/">DOOR</a> (operons)</li>
                                           </ul>
                                        <li>Precomputed homology searches with <a href="">RefSeq</a> and <a href="">SwissProt</a> databases for each protein</li>
                                        <li>Precomputed phylogenetic reconstructions of orthologous groups including the closest RefSeq homologs</li>
                                        <li>Pfam and COG phylogenetic profiles allowing to identify chlamydial domains/proteins rarely found in other prokaryotes (see for example <a href="{% url 'fam' "PF01823" "pfam" %}">PF01823</a>)</li>
                                        <li>Identification of conserved gene neighborhood among distantly related <i>Chlamydiae</i> and PVC (see for example the paper from <a href="https://www.ncbi.nlm.nih.gov/pubmed/9787636">Dandekar et al.</a> for the rationale behind this)</li>
                                        <li>Multiple tools allowing comparative analyses of <i>Chlamydiae</i> genomes</li>
                                        <li>References to scientific publications collected from multiple sources</li>

                                      </font>

                                    </ul>
                                  </div>
                                </div>
                              </div>

                              <div class="col-lg-6">


                                <div class="panel panel-default">
                                  <div class="panel-heading">
                                    <h3 class="panel-title">Latest News</h3>
                                  </div>
                                  <div class="panel-body" style="margin: 10px 10px 10px 10px">

                                    <font size="2">
                                        <p>
                                            <strong>December 02, 2019</strong>
                                        </p>
                                        <p>
                                            Updated references to litterature (reported in the "Pubmed" tab available on each locus page) using data from <a href="https://string-db.org/">STRING version 11</a>. 
                                        </p>
                                      <p>
                                          <strong>November 26, 2019</strong>
                                      </p>
                                      <p>
                                          New feature: Proteins associated to the same Pubmed entry (data from 
                                          <a href="http://papers.genomics.lbl.gov/cgi-bin/litSearch.cgi">PaperBlast</a>) 
                                          can be retrieved from the "PubMed" tab on each locus page. See for example papers 
                                          associated to <a href="{% url 'pmid' 7489 %}">the chaperone Scc2/CT576</a>
                                          and the <a href="{% url 'pmid_associations' 6 16159780 "PaperBlast" %}"> 9 
                                          <i>C. trachomatis</i> proteins associated to the publication from Fields et al.</a> 
                                          (<u>Analysis of putative <i>Chlamydia trachomatis</i> chaperones 
                                          Scc2 and Scc3 and their use in the identification of type III secretion substrates</u>, J Bacteriol. 2005).

                                      </p>
                                      <p>
                                          <strong>October 22, 2019</strong>
                                      </p>
                                      <p>
                                          Updated references to litterature (reported in the "Pubmed" tab available on each locus page) using the <a href="http://papers.genomics.lbl.gov/cgi-bin/litSearch.cgi">PaperBlast</a> database from September 25 2019. 
                                      </p>
                                      <strong>June 2019</strong>
                                    </p>
                                    <p>
                                      Release 2.0 of the database with 277 PVC genomes.
                                    </p>
                                    </font>

                                    </ul>
                                  </div>
                                </div>


                              </div>

                            </div>


                            <div class="row">
                              <div class="col-lg-12">
                                <div class="panel panel-default">
                                  <div class="panel-heading">
                                    <h3 class="panel-title">Citation</h3>
                                  </div>
                                  <div class="panel-body" style="margin: 10px 10px 10px 10px">

                                    <a href="https://academic.oup.com/nar/advance-article-abstract/doi/10.1093/nar/gkz924/5609527"> <strong> ChlamDB: a comparative genomics database of the phylum <i>Chlamydiae</i> and other members of the <i>Planctomycetes-Verrucomicrobiae-Chlamydiae</i> superphylum </strong></br>
                                    Trestan Pillonel, Florian Tagini, Claire Bertelli, Gilbert Greub </br>
                                    Nucleic Acids Research, 2019, https://doi.org/10.1093/nar/gkz924 </a>
                                  </div>
                                </div>
                              </div>
                            </div>


                        </div>
                        <div class="tab-pane" id="genomes">
                          <div class="col-lg-12 col-md-12 col-sm-12">
                           <div class="row">

                            <table class="display" id="genomes_table">
                                <thead>


                                    <tr>
                                        <th>Accession</th>
                                        <th>GC</th>
                                        <th>N proteins</th>
                                        <th>N contigs</th>
                                        <th>Size</th>
                                        <th>Percent non coding</th>
                                        <th>Organism</th>
                                        <th>Browse Online</th>
                                        <th>gbk</th>
                                        <th>faa</th>
                                        <th>fna</th>
                                        <th>ffn</th>
                                        <th>tab</th>

                                    </tr>
                                 </thead>
                                 <tbody>
                                 {% for values in genomes_data %}

                                    <tr>
                                        <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{values.0}}">{{values.0}}</a></td>
                                        <td>{{values.1}}</td>
                                        <td>{{values.2}}</td>
                                        <td>{{values.3}}</td>
                                        <td>{{values.4}}</td>
                                        <!--<td>{{values.5}}</td>-->
                                        <!--<td>{{values.6}}</td>-->
                                        <!--<td>{{values.7}}</td>-->
                                        <!--<td>{{values.8}}</td>-->
                                        <td>{{values.9}}</td>
                                        <td>{{values.13}}</td>
                                        <td><a href="{% url 'genome_annotation' values.0 %}">{{values.0}}</a></td>

                                        <td style="vertical-align:top" nowrap>
                                            <a  style="text-decoration:none;" href="{% url 'get_record' values.0 'gbk' %}">
                                            <span style="background-color:#DCFFF0;color:black;padding:1px;">GBK</span></a>
                                        </td>

                                        <td nowrap>
                                            <a  style="text-decoration:none;" href="{% url 'get_record' values.0 'faa' %}"">
                                            <span style="background-color:#DCFFF0;color:black;padding:1px;">FAA</span></a>
                                        </td>

                                        <td nowrap>
                                            <a  style="text-decoration:none;" href="{% url 'get_record' values.0 'fna' %}"">
                                            <span style="background-color:#DCFFF0;color:black;padding:1px;">FNA</span></a>
                                        </td>

                                        <td nowrap>
                                            <a  style="text-decoration:none;" href="{% url 'get_record' values.0 'ffn' %}">
                                            <span style="background-color:#DCFFF0;color:black;padding:1px;">FFN</span></a>
                                        </td>

                                        <td nowrap>
                                            <a  style="text-decoration:none;" href="{% url 'get_record' values.0 'gbk' %}">
                                            <span style="background-color:#DCFFF0;color:black;padding:1px;">TAB</span></a>

                                    </tr>

                                 {% endfor %}

                                </tbody>
                            </table>
                        </div><!-- row -->
                      </div> <!-- col -->
                    </div> <!-- tab0 -->
                        <div class="tab-pane" id="phylogeny">
                          <div class="row">
                            <div class="col-lg-12">
                              <div class="row">

                            <div class="panel panel-success" style="width:80% ; top: 200px; margin: 10px 10px 10px 10px">
                                <div class="panel-heading" style="width:100%">
                                    <h3 class="panel-title">Help</h3>
                                </div>
                                <p style="margin: 10px 10px 10px 10px; line-height: 160%">The phylogeny was build based on concatenated single copy orthologs with FastTree (default parameters). Nodes with black dots have support values lower than 1. 
                                The percentage of non-coding DNA of fragmented genomes was calculated based on data from contigs larger than 10 kilo bases. The completeness and contamination of each genome were estimated with <a href="https://ecogenomics.github.io/CheckM/">checkM</a> based on the identification of 104 "universal" marker genes.</p>
                                </p>
                            </div>




                                <a download="profile.svg" class="btn" href="{% static asset_path %}"><i class="fa fa-download"></i> Download SVG</a>
                                <a onclick='exportPNG("pyhlogeny", "phylogeny");' class="btn" id="png_button"><i class="fa fa-download"></i> Download PNG</a>
                              </div>
                              <div>
                                <object type="image/svg+xml" data="{% static asset_path %}" id="pyhlogeny" style="width:1200px"></object>
                              </div>
                            </div>
                          </div>
                        </div> <!- tab1 -->
                    </div>  <!--tab content -->
                    <!--<img src="{% static 'chlam_tree.png' %}" width="80%"-->
                </div> <!--col -->
              </div> <!--row -->
            </div> <!--page-content-wrapper -->
        </div> <!--wrapper -->
      </div> <!--container -->
    </div> <!--container -->

    <script>
      $(document).ready(function() {

      // Javascript to enable link to tab
      var url = document.location.toString();
      if (url.match('#')) {
          $('.nav-tabs a[href="#' + url.split('#')[1] + '"]').tab('show');
      } 

      // Change hash for page-reload
      $('.nav-tabs a').on('shown.bs.tab', function (e) {
          window.location.hash = e.target.hash;
      })

      $('#genomes_table').DataTable( {
      dom: 'Bfrtip',
      "paging":   true,
      "ordering": true,
      "info":     false,
      buttons: [
        {
            extend: 'excel',
            title: 'interaction_table'
        },
        {
            extend: 'csv',
            title: 'interaction_table'
        }
      ],
      } );
      } );
    $('.dropdown-toggle').dropdown()




    </script>




</body>
{% include "chlamdb/style_menu.html" %}


</html>
