<!DOCTYPE html>

<html>

<head>
  {% load custom_tags %}
  {% load static %}
  {% load crispy_forms_tags %}

  <script src="https://d3js.org/d3.v7.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/cgview/dist/cgview.min.js"></script>
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/cgview/dist/cgview.css">
  {% include "chlamdb/header.html" %}
</head>

<body>
  <div class="container-fluid" id="main_container">
    <div class="row">
      <div id="wrapper">
        <div id="page-content-wrapper">
          <div class="row">
            <div class="col-lg-12">
              {% include "chlamdb/menu.html" %}
              <p class="page-title">
                <b>Circos plots indicating the presence/absence of homologous proteins in one or multiple other genomes </b>
                <a href="https://zdb.readthedocs.io/en/latest/tutorial/website.html#genome-alignments" id="show-option" target="_blank"  title="Get a Circos plot to compare genomes (Targets) and visualize the presence/absence of homologous proteins according to the selected reference one (Circos reference)">
                  <i class="fab fa-info-circle " style="size: 5em;" ></i>
                </a>
              </p>
              {% block content %}
                {% csrf_token %}
                <span style="display:{{form_display}};">
                {% crispy form %}
                </span>
              {% endblock %}
            </div>

            {% if envoi %}
              <div class="row">
                <div class="col-lg-12">
                  <ul id="tabs" class="nav nav-tabs" data-tabs="tabs">
                    <br>
                    <div class="panel panel-success" style="width:100%; top: 200px ; margin: 10px 10px 10px 10px">
                      <div class="panel-heading" style="width:100%">
                        <h3 class="panel-title">Info about the analysis</h3>
                      </div>
                      <div style="padding-left:20px">
                        <br>

                        <ul id="circos_ul" style="list-style-type:disc;">
                          <li>
                            The first outer circle (<span style="color: rgb(79, 188, 221);"><b>light-blue</b></span>/<span style="color: red;"><b>red</b></span>) is made of barplots indicating the frequency of each orthogroup in the selected genomes.
                          </li>
                          <li>
                            In the second circle the contigs of the reference genome are displayed (<span style="color: orangered;"><b>orange</b></span>/<span style="color: rgb(81, 190, 166);"><b>aquamarine</b></span>)
                          </li>
                          <li>
                            The third and fourth circles represent the forward and reverse strands and show open reading frames from the reference genome:
                          </li>
                          {% if highlighted_ogs %}
                            <ol>
                              <li>
                                loci retrieved in extraction view are colored in <span style="color: rgb(200, 41, 248)"><b>pink</b></span>
                              </li>
                              <li>
                                others are colored in <span style="color: rgb(92, 92, 92)"><b>gray</b></span>
                              </li>
                            </ol>
                          {% else %}
                            <ol>
                              <li>
                                rRNA are colored in <span style="color: rgb(200, 41, 248)"><b>pink</b></span>
                              </li>
                              <li>
                                DNA is colored in <span style="color: rgb(92, 92, 92)"><b>grey</b></span>
                              </li>
                            </ol>
                          {% endif %}
                          <li>
                            Inner <span style="color: rgb(114, 22, 22);"><b>bordeaux</b></span>/<span style="color: rgb(79, 188, 221);"><b>light-blue</b></span> circle(s) indicate the presence(bordeaux scale)/absence(light blue) of homologous proteins in one or multiple other genomes
                          </li>
                          <li>
                            The inner circle (<span style="color: green;"><b>green</b></span>) indicate the GC content of each open reading frames of the reference genome
                          </li>
                          <br>
                          Click on features of the third circle (in <span style="color: rgb(92, 92, 92)"><b>gray</b></span> and <span style="color: rgb(200, 41, 248)"><b>pink</b></span>) to get detailed information about the corresponding locus.
                          <br>
                        </ul>
                      </div>
                    </div>
                  </ul>
                </div>
              </div>
            {% endif %}
            <div id='circos_plot'></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>

{% include "chlamdb/style_menu.html" %}

<script type="text/javascript">

$(document).ready(function() {
  // Create Viewer
  cgv = new CGV.Viewer('#circos_plot', {
    height: 500,
    width: 500
  });
  // Load data
  cgv.io.loadJSON({{circos_json| safe}});
  cgv.addTracks({
    name: 'GC Skew',
    position: 'inside',
    dataType: 'feature',
    dataMethod: 'sequence',
    // e.g. In this example, GC Skew plot will be extracted from the sequence.
    dataKeys: 'gc-skew',
    // By default, the step and window values are calculated based on the sequence length,
    // but they can be overridden here.
    dataOptions: {
      step: 1,
      window: 100
    }
  });
  cgv.addTracks({
    name: 'GC Content',
    position: 'inside',
    dataType: 'feature',
    dataMethod: 'sequence',
    // e.g. In this example, GC Content plot will be extracted from the sequence.
    dataKeys: 'gc-content',
    // By default, the step and window values are calculated based on the sequence length,
    // but they can be overridden here.
    dataOptions: {
      step: 1,
      window: 100
    }
  });
  cgv.on('click', (event) => {
    if (event.elementType === 'feature') {
      if (['orthogroups', 'genes'].includes(event.slot.track.name)) {
        window.open(`/locusx/${event.element.name}`)
      }
    }
  });
  cgv.draw()
});

</script>

</html>
