
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Website tutorial &#8212; ChlamDB 2.0 (June 2019) documentation</title>
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script type="text/javascript" src="../_static/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/js/custom.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Methods" href="../methods/annotation.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../methods/annotation.html" title="Methods"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ChlamDB 2.0 (June 2019) documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="website-tutorial">
<h1>Website tutorial<a class="headerlink" href="#website-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This page is under construction. Contact Trestan Pillonel (<a class="reference external" href="mailto:trestan&#46;pillonel&#37;&#52;&#48;chuv&#46;ch">trestan<span>&#46;</span>pillonel<span>&#64;</span>chuv<span>&#46;</span>ch</a>) if you have any question or suggestion regarding the website or its documentation.</p>
</div>
<div class="section" id="search-for-a-protein-of-interest">
<h2>Search for a protein of interest<a class="headerlink" href="#search-for-a-protein-of-interest" title="Permalink to this headline">¶</a></h2>
<p>The following protein accessions are recognised:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 60%" />
<col style="width: 40%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Genbank locus tag</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=wcw_1594">wcw_1594</a></p></td>
</tr>
<tr class="row-odd"><td><p>Genbank protein accession</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=ADI38940.1">ADI38940.1</a></p></td>
</tr>
<tr class="row-even"><td><p>RefSeq locus tag</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=WCW_RS07680">WCW_RS07680</a></p></td>
</tr>
<tr class="row-odd"><td><p>RefSeq protein accession</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=WP_013182646.1">WP_013182646.1</a></p></td>
</tr>
<tr class="row-even"><td><p>UniParc accession</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=UPI0001D5C1DD">UPI0001D5C1DD</a></p></td>
</tr>
<tr class="row-odd"><td><p>UniProtKB-Accession</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=D6YS95">D6YS95</a></p></td>
</tr>
<tr class="row-even"><td><p>UniProtKB-ID</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=D6YS95_WADCW">D6YS95_WADCW</a></p></td>
</tr>
</tbody>
</table>
<p></p>
<p></p>
<p>It’s also possible to seach for a gene or product name:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=mreb">mreb</a></p></li>
<li><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=mreb+Waddlia">mreb Waddlia</a></p></li>
<li><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=secretion+system">secretion system</a></p></li>
</ul>
</div></blockquote>
<p></p>
<p>Please note that the search is performed in coding sequence annotations but also in <code class="docutils literal notranslate"><span class="pre">EC</span></code>, <code class="docutils literal notranslate"><span class="pre">Kegg</span> <span class="pre">Orthologs</span></code>, <code class="docutils literal notranslate"><span class="pre">Interpro</span></code> and <code class="docutils literal notranslate"><span class="pre">Kegg</span>
<span class="pre">Pathways/Modules</span></code> description fields. The results (if matches were found) are reported in separate tabs (Figure 1).</p>
<p></p>
<div class="align-center figure align-default" id="id1">
<a class="reference internal image-reference" href="../_images/search.png"><img alt="../_images/search.png" src="../_images/search.png" style="width: 90%;" /></a>
<p class="caption"><span class="caption-text"><strong>Figure 1:</strong> Search result for <code class="docutils literal notranslate"><span class="pre">secretion</span> <span class="pre">system</span></code>. Note the presence of multiple tabs with search results in
coding sequence annotations (“<strong>locus</strong> tab”), but also in <strong>EC</strong>, <strong>Kegg Orthologs</strong> (KO),
<strong>Interpro</strong> and <strong>Kegg Pathways</strong>/<strong>Modules</strong> descriptions.</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
<p>It’s also possible to browse genomes tables from links listed in the <a class="reference external" href="https://chlamdb.ch/#genomes">ChlamDB home page</a> (column: Browse online)</p>
</div>
<div class="section" id="species-phylogeny">
<h2>Species phylogeny<a class="headerlink" href="#species-phylogeny" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="orthogroup-phylogenies">
<h2>Orthogroup phylogenies<a class="headerlink" href="#orthogroup-phylogenies" title="Permalink to this headline">¶</a></h2>
<p>“Putative interactors were predicted in-house from genomic data alone using phylogenetic profiling and investigation of conserved gene neighborhood (see online methods). Phylogenetic profile similarity (pattern of presence/absence of orthologs within the PVC superphylum) was calculated using Euclidean and Jaccard distances. Conserved neighbors were identified by identifying orthologs encoded less than 20 kilobases apart in genomes from different species of the PVC superphylum (Figure 1.H). See (19) and (20) for the rationale justifying use of those two approaches.”</p>
</div>
<div class="section" id="search-using-cog-pfam-interpro-or-kegg-accessions">
<h2>Search using COG, Pfam, Interpro or KEGG accessions<a class="headerlink" href="#search-using-cog-pfam-interpro-or-kegg-accessions" title="Permalink to this headline">¶</a></h2>
<p>Accessions from <code class="docutils literal notranslate"><span class="pre">KEGG</span></code>, <code class="docutils literal notranslate"><span class="pre">COG</span></code>, <code class="docutils literal notranslate"><span class="pre">Pfam</span></code> and <code class="docutils literal notranslate"><span class="pre">InterPro</span></code> can also be searched.
The result page will report a summary of the entry,  the list of proteins annotated
with this entry as well as a figure showing the presence/absence of this annotation
in all genomes included in the database (Figure 2).
</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 21%" />
<col style="width: 12%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Accession type</p></th>
<th class="head"><p>Example</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>KEGG ortholog</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=K00844">K00844</a></p></td>
<td><p>hexokinase [EC:2.7.1.1]</p></td>
</tr>
<tr class="row-odd"><td><p>COG</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=COG0333">COG0333</a></p></td>
<td><p>Ribosomal protein L32</p></td>
</tr>
<tr class="row-even"><td><p>PFAM</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=PF06723">PF06723</a></p></td>
<td><p>MreB/Mbl protein</p></td>
</tr>
<tr class="row-odd"><td><p>InterPro</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=IPR004753">IPR004753</a></p></td>
<td><p>Cell shape determining protein MreB</p></td>
</tr>
<tr class="row-even"><td><p>KEGG modules</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=M00023">M00023</a></p></td>
<td><p>Amino acid metabolism</p></td>
</tr>
<tr class="row-odd"><td><p>Kegg pathways</p></td>
<td><p><a class="reference external" href="https://chlamdb.ch/locusx?accession=map00400">map00400</a></p></td>
<td><p>Phenylalanine, tyrosine and tryptophan biosynthesis</p></td>
</tr>
</tbody>
</table>
<p></p>
<p></p>
<div class="align-center figure align-default" id="id2">
<a class="reference internal image-reference" href="../_images/K01902_profile.svg"><img alt="../_images/K01902_profile.svg" src="../_images/K01902_profile.svg" width="60%" /></a>
<p class="caption"><span class="caption-text"><strong>Figure 2:</strong> See the <a class="reference external" href="https://chlamdb.ch/locusx?accession=K01902#tab3">complete profile online</a>.</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
<p></p>
<p></p>
<div class="align-center figure align-default">
<a class="reference internal image-reference" href="../_images/TCA_MAP.svg"><img alt="../_images/TCA_MAP.svg" src="../_images/TCA_MAP.svg" width="90%" /></a>
</div>
</div>
<div class="section" id="taxonomic-profile-of-cogs-and-pfam-domains">
<h2>Taxonomic profile of COGs and Pfam domains<a class="headerlink" href="#taxonomic-profile-of-cogs-and-pfam-domains" title="Permalink to this headline">¶</a></h2>
<div class="align-center figure align-default">
<a class="reference internal image-reference" href="../_images/PF08486_phylum_profile1.svg"><img alt="../_images/PF08486_phylum_profile1.svg" src="../_images/PF08486_phylum_profile1.svg" width="600px" /></a>
</div>
</div>
<div class="section" id="blast-interface">
<h2>BLAST interface<a class="headerlink" href="#blast-interface" title="Permalink to this headline">¶</a></h2>
<p>A BLAST interface is also available for homology search:</p>
<div class="align-center figure align-default" id="id3">
<img alt="../_images/screenshot_blast.png" src="../_images/screenshot_blast.png" />
<p class="caption"><span class="caption-text">Figure 1: Simplified annotation workflow.</span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="protein-annotation-view">
<h2>Protein annotation view<a class="headerlink" href="#protein-annotation-view" title="Permalink to this headline">¶</a></h2>
<div class="align-center figure align-default" id="id4">
<img alt="../_images/locus_page.svg" src="../_images/locus_page.svg" /><p class="caption"><span class="caption-text">Figure 1: Simplified annotation workflow.</span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="orthogroup-annotation-summary">
<h2>Orthogroup annotation summary<a class="headerlink" href="#orthogroup-annotation-summary" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="alignments-of-target-genomic-regions-from-two-or-more-genomes">
<h2>Alignments of target genomic regions (from two or more genomes)<a class="headerlink" href="#alignments-of-target-genomic-regions-from-two-or-more-genomes" title="Permalink to this headline">¶</a></h2>
<div class="align-center figure align-default" id="id5">
<img alt="../_images/region_align.svg" src="../_images/region_align.svg" /><p class="caption"><span class="caption-text">Figure 1: Simplified annotation workflow.</span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="whole-genomes-alignments-interactive-circular-plots">
<h2>Whole genomes alignments: interactive circular plots<a class="headerlink" href="#whole-genomes-alignments-interactive-circular-plots" title="Permalink to this headline">¶</a></h2>
<div class="align-center figure align-default" id="id6">
<img alt="../_images/circos_interactive.png" src="../_images/circos_interactive.png" />
<p class="caption"><span class="caption-text">Figure 1: Simplified annotation workflow.</span><a class="headerlink" href="#id6" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="comparing-the-content-of-two-or-more-genomes">
<h2>comparing the content of two or more genomes<a class="headerlink" href="#comparing-the-content-of-two-or-more-genomes" title="Permalink to this headline">¶</a></h2>
<div class="section" id="overview">
<h3>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h3>
<div class="align-center figure align-default" id="id7">
<img alt="../_images/extract_orthogroup_page.png" src="../_images/extract_orthogroup_page.png" />
<p class="caption"><span class="caption-text">Figure 1: Simplified annotation workflow.</span><a class="headerlink" href="#id7" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="venn-diagrams-to-compare-genome-content">
<h3>Venn diagrams to compare genome content<a class="headerlink" href="#venn-diagrams-to-compare-genome-content" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="identification-of-the-largest-protein-families-in-one-or-multiple-genomes">
<h3>Identification of the largest protein families in one or multiple genomes<a class="headerlink" href="#identification-of-the-largest-protein-families-in-one-or-multiple-genomes" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="proteomes-heatmaps">
<h3>Proteomes heatmaps<a class="headerlink" href="#proteomes-heatmaps" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="pan-core-genome-plots">
<h3>Pan/Core genome plots<a class="headerlink" href="#pan-core-genome-plots" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="comparison-of-cog-pfam-ko-and-interpro-annotations-at-the-genome-scale">
<h2>Comparison of COG, Pfam, KO and InterPro annotations at the genome scale<a class="headerlink" href="#comparison-of-cog-pfam-ko-and-interpro-annotations-at-the-genome-scale" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li><p>identification of conserved or clade specific domains/COGs,EC numbers</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="get-the-list-of-pfam-domains-specific-to-one-or-a-set-of-genomes">
<h2>Get the list of Pfam domains specific to one or a set of genomes<a class="headerlink" href="#get-the-list-of-pfam-domains-specific-to-one-or-a-set-of-genomes" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="metabolism">
<h2>Metabolism<a class="headerlink" href="#metabolism" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Website tutorial</a><ul>
<li><a class="reference internal" href="#search-for-a-protein-of-interest">Search for a protein of interest</a></li>
<li><a class="reference internal" href="#species-phylogeny">Species phylogeny</a></li>
<li><a class="reference internal" href="#orthogroup-phylogenies">Orthogroup phylogenies</a></li>
<li><a class="reference internal" href="#search-using-cog-pfam-interpro-or-kegg-accessions">Search using COG, Pfam, Interpro or KEGG accessions</a></li>
<li><a class="reference internal" href="#taxonomic-profile-of-cogs-and-pfam-domains">Taxonomic profile of COGs and Pfam domains</a></li>
<li><a class="reference internal" href="#blast-interface">BLAST interface</a></li>
<li><a class="reference internal" href="#protein-annotation-view">Protein annotation view</a></li>
<li><a class="reference internal" href="#orthogroup-annotation-summary">Orthogroup annotation summary</a></li>
<li><a class="reference internal" href="#alignments-of-target-genomic-regions-from-two-or-more-genomes">Alignments of target genomic regions (from two or more genomes)</a></li>
<li><a class="reference internal" href="#whole-genomes-alignments-interactive-circular-plots">Whole genomes alignments: interactive circular plots</a></li>
<li><a class="reference internal" href="#comparing-the-content-of-two-or-more-genomes">comparing the content of two or more genomes</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#venn-diagrams-to-compare-genome-content">Venn diagrams to compare genome content</a></li>
<li><a class="reference internal" href="#identification-of-the-largest-protein-families-in-one-or-multiple-genomes">Identification of the largest protein families in one or multiple genomes</a></li>
<li><a class="reference internal" href="#proteomes-heatmaps">Proteomes heatmaps</a></li>
<li><a class="reference internal" href="#pan-core-genome-plots">Pan/Core genome plots</a></li>
</ul>
</li>
<li><a class="reference internal" href="#comparison-of-cog-pfam-ko-and-interpro-annotations-at-the-genome-scale">Comparison of COG, Pfam, KO and InterPro annotations at the genome scale</a></li>
<li><a class="reference internal" href="#get-the-list-of-pfam-domains-specific-to-one-or-a-set-of-genomes">Get the list of Pfam domains specific to one or a set of genomes</a></li>
<li><a class="reference internal" href="#metabolism">Metabolism</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../methods/annotation.html"
                        title="previous chapter">Methods</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorial/website.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../methods/annotation.html" title="Methods"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ChlamDB 2.0 (June 2019) documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Trestan Pillonel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.
    </div>
  </body>
</html>